### PR TITLE
Ordinal election

### PR DESCRIPTION
**Pull request by Wolokin**

### CODE REVIEW

File: votepy/ordinal_election.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
+from typing import Iterable, Sequence
+
+
+class OrdinalBallot(list):
+    def __init__(self, ordering: Iterable, mapping=None):
+        super().__init__(ordering)
+        
+        self.mapping = mapping
+        
+        for candidate in ordering:
```
Comment (Janix4000) [inline votepy/ordinal_election.py]: If you want to allow all types of iterable objects, you need to ensure only one iteration over it. For example, the generator returned by `map` will be depleted after line 6.

Comment (Janix4000) [reply votepy/ordinal_election.py]: You should iterate over `self` or shadow variable `ordering`

File: votepy/ordinal_election.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+from typing import Iterable, Sequence
+
+
+class OrdinalBallot(list):
+    def __init__(self, ordering: Iterable, mapping=None):
```
Comment (Janix4000) [inline votepy/ordinal_election.py]: You can specify the type of the objects in the `Iterable` using `[]`, e.g. `Iterable[int]`

Comment (Janix4000) [inline votepy/ordinal_election.py]: What is the type of `mapping`?

File: votepy/ordinal_election.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+                raise ValueError(f"Only non-negative integers are valid candidates, value {candidate} was supplied")
+            
+    def __str__(self) -> str:
+        if self.mapping is not None:
+            return str([self.mapping[candidate] for candidate in self])
+        return str([self])
+    
+    
+class OrdinalElection(list):
+    def __init__(self, preference_orders: list, mapping=None):
```
Comment (Janix4000) [inline votepy/ordinal_election.py]: Why `list` over `Iterable`/`Sequence`? Inconsistent typing.

File: votepy/ordinal_election.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        super().__init__()
+        
+        self.mapping = mapping
+        self.candidates = set() if len(preference_orders) == 0 else set(preference_orders[0])
```
Comment (Janix4000) [inline votepy/ordinal_election.py]: `if preference_orders` is more pythonic

File: votepy/ordinal_election.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        
+        for preference in preference_orders:
+            if len(self.candidates) != len(preference):
+                raise ValueError(f"Preference orders are of different lengths")
+            for candidate in preference:
+                if candidate not in self.candidates:
+                    raise ValueError(f"Candidate sets in preference orders differ: candidate {candidate} is not present in every preference order")
+            self.append(OrdinalBallot(preference, mapping))
+            
+    def update_mapping(self, mapping: Sequence) -> None:
```
Comment (Janix4000) [inline votepy/ordinal_election.py:41]: Can `mapping` be one of all possible Python sequences?

File: votepy/ordinal_election.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        self.mapping = mapping
+        
+    def __str__(self) -> str:
+        return "\n".join(map(lambda x: str(x), self))
```
Comment (Janix4000) [inline votepy/ordinal_election.py]: You can just use `str` in the first `map` argument.

**Review by Janix4000 — CHANGES_REQUESTED** [review]
Typing issues

**Comment by Janix4000** [comment]
[Implementing class structure](https://trello.com/c/irhM1I0d/15-implementing-class-structure)

File: votepy/ordinal_election.py (hunk 1 above)
Comment (Wolokin) [reply votepy/ordinal_election.py]: True, fixed

File: votepy/ordinal_election.py (hunk 2 above)
Comment (Wolokin) [reply votepy/ordinal_election.py]: Ok, thanks

File: votepy/ordinal_election.py (hunk 4 above)
Comment (Wolokin) [reply votepy/ordinal_election.py]: True, fixed

File: votepy/ordinal_election.py (hunk 5 above)
Comment (Wolokin) [reply votepy/ordinal_election.py:41]: Mapping is supposed to only have __getitem__ method, and I found Sequence to be the most primitive type supporting it. As a side note it seems that most typing classes from the 'typing' lib are deprecated, so I will migrate to collections.abc.Sequence.

File: votepy/ordinal_election.py (hunk 6 above)
Comment (Wolokin) [reply votepy/ordinal_election.py]: Yeah lol

File: votepy/ordinal_election.py (hunk 2 above)
Comment (Wolokin) [reply votepy/ordinal_election.py]: Missed that, it's the same as in 'update_mapping'. Fixed

File: votepy/ordinal_election.py (hunk 3 above)
Comment (Wolokin) [reply votepy/ordinal_election.py]: No reason, replaced with Iterable

File: votepy/ordinal_election.py (hunk 5 above)
Comment (Janix4000) [reply votepy/ordinal_election.py:41]: Ok

**Review by Janix4000 — APPROVED** [review]

**Pull request merged**